backend/storage-service/src/jvmMain/kotlin/file/services/SyncthingClient.kt

```diff
@@ -0,0 +1,304 @@
+    val devices: List<SyncthingDevice> = emptyList(),
+    val folders: List<SyncthingFolder> = emptyList(),
+    val gui: SyncthingGui,
+    val ldap: SyncthingLdap,
+    val options: SyncthingOptions,
+    val remoteIgnoredDevices: List<SyncthingRemoteIgnoredDevices> = emptyList(),
+    val version: Int = 35
+)
+
+@Serializable
+data class SyncthingFolder(
+    val autoNormalize: Boolean = true,
+    val blockPullOrder: String = "standard",
+    val caseSensitiveFS: Boolean = false,
+    val copiers: Int = 0,
+    val copyOwnershipFromParent: Boolean = false,
+    val copyRangeMethod: String = "standard",
+    val devices: List<SyncthingFolderDevice> = emptyList(),
+    val disableFsync: Boolean = false,
+    val disableSparseFiles: Boolean = false,
+    val disableTempIndexes: Boolean = false,
+    val filesystemType: String = "basic",
+    val fsWatcherDelayS: Int = 10,
+    val fsWatcherEnabled: Boolean = true,
+    val hashers: Int = 0,
+    val id: String = "",
+    val ignoreDelete: Boolean = false,
+    val ignorePerms: Boolean = false,
+    val junctionsAsDirs: Boolean = false,
+    val label: String = "",
+    val markerName: String = ".stfolder",
+    val maxConcurrentWrites: Int = 2,
+    val maxConflicts: Int = 10,
+    val minDiskFree: MinDiskFree = MinDiskFree(),
+    val modTimeWindowS: Int = 0,
+    val order: String = "random",
+    val path: String = "",
+    val paused: Boolean = false,
+    val pullerMaxPendingKiB: Int = 0,
+    val pullerPauseS: Int = 0,
+    val rescanIntervalS: Int = 3600,
+    val scanProgressIntervalS: Int = 0,
+    val type: String = SynchronizationType.SEND_RECEIVE.syncthingValue,
+    val versioning: Versioning = Versioning(),
+    val weakHashThresholdPct: Int = 25
+)
+
+@Serializable
+data class MinDiskFree(
+    val unit: String = "%",
+    val value: Int = 1
+)
+
+@Serializable
+data class Versioning(
+    val cleanupIntervalS: Int = 0,
+    val fsPath: String = "",
+    val fsType: String = "basic",
+    val params: VersioningParams = Unit,
+    val type: String = ""
+)
+
+typealias VersioningParams = Unit
+
+
+@Serializable
+data class SyncthingFolderDevice(
+    val deviceID: String,
+    val encryptionPassword: String = "",
+    val introducedBy: String = ""
+)
+
+class SyncthingClient(
+    val config: SynchronizationConfiguration,
+    val db: DBContext,
+    val fsPath: String
+) {
+    private val httpClient = HttpClient(CIO) {
+        expectSuccess = false
+    }
+
+    private suspend fun readConfig(device: LocalSyncthingDevice): SyncthingConfig {
+        val requestPath = "http://" + device.hostname + "/rest/config"
+        val resp = httpClient.get<HttpResponse>(requestPath) {
+            headers {
+                append("X-API-Key", device.apiKey)
+            }
+        }
+
+        val config = resp.content.toByteArray().toString(Charsets.UTF_8)
+        return defaultMapper.decodeFromString(config)
+    }
+
+    suspend fun writeConfig() {
```

Comment:
This function is called every time any change is made. Changes are made as a result of RPCs caused by a user action. This means that you have _many_ different threads hitting this call, yet there is currently no locking to prevent multiple instances of this function going off at the same time. You need to add some kind of locking to this function (e.g. `Mutex`).